contracts/MarketCapSqrtController.sol

```diff
@@ -612,18 +612,22 @@ contract MarketCapSqrtController is MarketCapSortedTokenCategories {
 /* ==========  Internal Pool Utility Functions  ========== */
 
   /**
-   * @dev Estimate the total value of a pool by taking its first token's
-   * "virtual balance" (balance * (totalWeight/weight)) and multiplying
-   * by that token's average ether price from UniSwap.
+   * @dev Estimate the total value of a pool by taking the sum of
+   * TWAP values of the pool's balance in each token it has bound.
    */
-  function _estimatePoolValue(IIndexPool pool) internal view returns (uint144) {
-    (address token, uint256 value) = pool.extrapolatePoolValueFromToken();
-    return oracle.computeAverageEthForTokens(
-      token,
-      value,
+  function _estimatePoolValue(IIndexPool pool) internal view returns (uint256 totalValue) {
+    address[] memory tokens = pool.getCurrentTokens();
+    uint256 len = tokens.length;
+    uint256[] memory balances = new uint256[](len);
+    for (uint256 i; i < len; i++) balances[i] = IERC20(tokens[i]).balanceOf(address(pool));
+    uint144[] memory ethValues = oracle.computeAverageEthForTokens(
+      tokens,
+      balances,
       SHORT_TWAP_MIN_TIME_ELAPSED,
       SHORT_TWAP_MAX_TIME_ELAPSED
     );
+    // Safe math is not needed because we are taking the sum of an array of uint144s as a uint256.
```

Comment:
Guessing this is safe because max uint144 is very very far from max uint256?

Reply:
Correct, the array would have to have 5.192296858534828e+33 tokens which each had a value of 2**144-1 for that to overflow.